feat(js): adds elgg/popup AMD module

hypeJunction opened the pull request:
Converts JS popup to an AMD module subsequently allowing plugins to
invoke and close popup programmically.
Popups can now be bound with .elgg-popup class in addition to rel=popup.
Popup positioning details can be passed with data- attributes of the trigger.

hypeJunction committed:
feature(js): adds elgg/popup AMD module

Converts JS popup to an AMD module subsequently allowing plugins to
manipulate popups programmatically.
Popups can now be bound with .elgg-popup class and data-href attribute in
addition to rel=popup and href.
Popup positioning details can be passed with data-position of the trigger.
Triggers additional 'open' and 'close' jQuery events that can be used to
manipulate the popup element and its contents

hypeJunction commented:
Let's see if you think this will work in 2.x. I think it's BC, though I refactored the code for opening and closing. Once this is in:

- [ ] Move user-hover menu logic to use the `elgg/popup` module to get rid of code duplication
- [ ] In 3.0, do not append the module to `body` - breaking change that will probably require z-index changes.

hypeJunction commented on docs/guides/javascript.rst:
```diff
@@ -434,6 +434,59 @@ The ``elgg/spinner`` module can be used to create an Ajax loading indicator fixe
+      'class' => 'elgg-popup elgg-button elgg-button-dropdown',
+      'href' => '#popup-element',
+      'text' => 'Open popup above',
+      'data-my' => 'center bottom',
```
Should we make it `data-position` with json?

mrclay replied:
:+1:

hypeJunction commented:
Screenshot from the theme sandbox
![popup-button](https://cloud.githubusercontent.com/assets/1202761/13577881/14cb47ae-e495-11e5-9368-ec026a233a4e.png)

mrclay commented on js/lib/ui.js:
```diff
@@ -18,7 +18,9 @@ elgg.ui.init = function () {
 
 	$(document).on('click', '[rel=toggle]', elgg.ui.toggles);
 
-	$(document).on('click', '[rel=popup]', elgg.ui.popupOpen);
```
Old code handles HTML added by Ajax. For BC we should do that as well.

hypeJunction replied:
This works with AJAX. The bindings are proxied.

hypeJunction committed:
feature(js): user hover menu now uses elgg/popup module

Open and close behaviour of the user hover menu is now consistent with other
overlays visible on the page. Positioning is now taken care of by the jQuery UI
position rather than customized maths.

hypeJunction commented:
Improved the guides and added some inline docs.
Now uses `data-position`, and also accepts `data-href` if `href` is not present.
User hover menu now uses the `elgg/popup` module.
Added jQuery events we can use to populate the popup dynamically.

@juho-jaakkola Can you take a look, I had notifier in mind when working on this.

mrclay commented on js/lib/ui.js:
```diff
@@ -100,87 +105,30 @@ elgg.ui.toggles = function(event) {
  *
  * @param {Object} event
  * @return void
+ * @deprecated 2.1
```
2.2

mrclay commented on views/default/elgg/popup.js:
```diff
@@ -0,0 +1,172 @@
+			});
+
+			// Adding this magic so that popups with fixed position stick to their parent element
+			$(window).on('scroll resize', function () {
```
Should we throttle this event by 100ms? Or can we detect fixed position to make sure we have to do this?

mrclay replied:
Is this required for the AMD migration or is this a bugfix we could move to another PR?

hypeJunction replied:
I believe currently it remains in the viewport when you scroll, so it's a UX bug. I will see if I can migrate with minimal changes and then fluff it up in additional PRs

hypeJunction commented:
Replaced by #9599
